server/controllers/firestore.py

```diff
@@ -12,59 +12,110 @@ def __init__(self, project_id=None, oauth2_credentials=None):
         )
         client = db.collection('students')
         self.metadata = client.document('metadata')
+        self.testimonials = db.collection('testimonials')
+        self.testimonials_metadata = self.testimonials.document('metadata')
         super(FirestoreDriver, self).__init__(client)
 
-    def add(self, data):
+    def _add(
+        self,
+        data,
+        *,
+        validation_fields,
+        client,
+        metadata,
+        check_unique=False,
+        check_exists=False,
+        validation_client=None
+    ):
         found_item = None
-        try:
-            found_item = self._find({'email': data['email']})
-            found_item = self._find({'student_number': data['student_number']})
-        except StopIteration:
-            pass
-
-        if found_item is not None:
-            return None
-
-        _, doc_ref = self.database_client.add(data)
-        self.metadata.update({
+        for field in validation_fields:
+            try:
+                found_item = self._find({field, data[field]}, validation_client)
+            except StopIteration:
+                pass
+
+            if check_unique and found_item is not None:
+                return None
+            if check_exists and found_item is None:
+                return None
+
+        _, doc_ref = client.add(data)
+        metadata.update({
             'total_documents': firestore.Increment(1)
         })
+
         return doc_ref.id
 
-    def _find(self, identifier):
+    def _find(self, identifier, query=None):
+        if query is None:
+            query = self.database_client
+
         search_fields = identifier.keys()
-        query = self.database_client
         for field in search_fields:
             query = query.where(field, '==', identifier[field])
         document = next(query.stream())
         return document
 
+    def _verify(self, id_str, *, client, metadata):
+        document = client.document(id_str)
+
+        if document.get().to_dict()['verified'] is True:
```

Comment:
```suggestion
        if document.get().to_dict()['verified']:
```
Unless this attribute might contain other truthy values?

Reply:
I left it like that on purpose to give a bit of a type hint that the value stored isn't just truthy but boolean

Reply:
That makes sense!  In general I might just make a comment for that sort of thing, but it's by no means important, so all good!